flexible mlp policies (shared and non-shared)

erniejunior opened the pull request:
This PR solves part of #91. The `LstmPolicy` is untouched but `FeedForwardPolicy` supports the new format to specify network architectures as discussed in #91.
This PR also includes:
- [x] documentation in code
- [x] documentation in the user guide
- [x] (visual) examples in the user guide
- [x] warnings when the old `layers` parameter is used
- [x] update of all the documentation to the new `net_arch` format instead of the old `layers` format
- [x] update of all test code to use the new `net_arch` format

hill-a commented on stable_baselines/common/policies.py:
```diff
@@ -236,41 +295,43 @@ class FeedForwardPolicy(ActorCriticPolicy):
+            print("Warning: usage of the `layers` parameter is deprecated! Use net_arch instead "
+                  "(it has a different semantics though).")
+            if net_arch is not None:
+                print("Warning: the new `net_arch` parameter overrides the `layers` parameter!")
```
Use the [python warning system](https://docs.python.org/3/library/warnings.html), it will helps devs when they are trying to control their stdout and stderr:
```python
import warnings

warnings.warn("the new `net_arch` parameter overrides the `layers` parameter!", DeprecationWarning)
```

erniejunior commented:
@hill-a Thanks for the feedback. I implemented the warnings, hopefully fixed the bugs I introduced by renaming `processed_x` to `processed_obs` and fixed some formatting issues. Lets see what the CI has to say about it ...

erniejunior commented:
When loading a model stored with the old version (two non-shared 128 layers), I get the following error:
```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 1628, in _create_c_op
    c_op = c_api.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimension 0 in both shapes must be equal, but are 128 and 18. Shapes are [128,128] and [18,128]. for 'Assign_2' (op: 'Assign') with input shapes: [128,128], [18,128].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/m/misc/pycharm-2018.1.3/helpers/pydev/pydevd.py", line 1664, in <module>
    main()
  File "/home/m/misc/pycharm-2018.1.3/helpers/pydev/pydevd.py", line 1658, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/home/m/misc/pycharm-2018.1.3/helpers/pydev/pydevd.py", line 1068, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/home/m/misc/pycharm-2018.1.3/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/m/volatile/Master Thesis/experiment_code/analyze_results.py", line 170, in <module>
    figures.add(make_position_bias_plot(trials), "position_bias")
  File "/home/m/volatile/Master Thesis/experiment_code/analyze_results.py", line 52, in make_position_bias_plot
    model = restore_model(trials, trial)
  File "/home/m/volatile/Master Thesis/experiment_code/analyze_results.py", line 39, in restore_model
    model = PPO2.load(bio)
  File "/home/m/volatile/Master Thesis/stable-baselines/stable_baselines/common/base_class.py", line 373, in load
    restores.append(param.assign(loaded_p))
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/variables.py", line 1718, in assign
    name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/state_ops.py", line 221, in assign
    validate_shape=validate_shape)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/gen_state_ops.py", line 61, in assign
    use_locking=use_locking, name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 1792, in __init__
    control_input_ops)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 1631, in _create_c_op
    raise ValueError(str(e))
ValueError: Dimension 0 in both shapes must be equal, but are 128 and 18. Shapes are [128,128] and [18,128]. for 'Assign_2' (op: 'Assign') with input shapes: [128,128], [18,128].
```
Do you have an idea what could cause this bug? I tried to put the same tf variables in the same scopes. Right now making this PR backwards compatible in regard to loading older models is not very high on my list of priorities. If nobody else knows a solution I would be happy to at least put in a warning when an old model is loaded so users know what is going wrong.

hill-a commented:
Well, i've just figured out why this happens. Turns out Tensorflow like to register weights by order of creation, as the names of identical operations are ordered sequentially (https://www.tensorflow.org/guide/graphs#naming_operations).

So instead of building the layers at the same time:
```python
processed_x = tf.layers.flatten(self.processed_x)
pi_h = processed_x
vf_h = processed_x
for i, layer_size in enumerate(layers):
    pi_h = activ(linear(pi_h, 'pi_fc' + str(i), n_hidden=layer_size, init_scale=np.sqrt(2)))
    vf_h = activ(linear(vf_h, 'vf_fc' + str(i), n_hidden=layer_size, init_scale=np.sqrt(2)))
value_fn = linear(vf_h, 'vf', 1)
pi_latent = pi_h
vf_latent = vf_h
```

this branch creates then one at a time:
```python
# Build part of the network that exclusively belongs to the policy network
latent_policy = latent
for idx, layer_size in enumerate(policy_only_layers):
    assert isinstance(layer_size, int)
    latent_policy = act_fun(linear(latent_policy, "pi_fc{}".format(idx), layer_size, init_scale=np.sqrt(2)))

# Build part of the network that exclusively belongs to the value network
latent_value = latent
for idx, layer_size in enumerate(value_only_layers):
    assert isinstance(layer_size, int)
    latent_value = act_fun(linear(latent_value, "vf_fc{}".format(idx), layer_size, init_scale=np.sqrt(2)))
```

causing the weights to be loaded out of order. I'll try and hack up clean-ish solution.

hill-a commented:
Honestly, the only good way to fix this issue, it by fixing the save and load methods, so that loading by name can be done. 

In the mean time, it's annoying and ugly, but you can do this:
```python
latent_policy = latent
latent_value = latent
for idx in range(max(len(policy_only_layers), len(value_only_layers))):
    if len(policy_only_layers) < idx:
        assert isinstance(policy_only_layers[idx], int)
        latent_policy = act_fun(linear(latent_policy, "pi_fc{}".format(idx), 
                                       policy_only_layers[idx], init_scale=np.sqrt(2)))
    if len(value_only_layers) < idx:
        assert isinstance(value_only_layers[idx], int)
        latent_value = act_fun(linear(latent_value, "vf_fc{}".format(idx), 
                                      value_only_layers[idx], init_scale=np.sqrt(2)))
```

araffin commented on docs/guide/custom_policy.rst:
```diff
@@ -65,6 +67,62 @@ You can also registered your policy, to help with code simplicity: you can refer
 
 
+
+The net_arch parameter of `FeedForwardPolicy` allows to specify the amount and size of the hidden layers and how many
```
```suggestion
The net_arch parameter of ``FeedForwardPolicy`` allows to specify the amount and size of the hidden layers and how many
```

araffin commented on stable_baselines/common/policies.py:
```diff
@@ -8,6 +8,8 @@
 from stable_baselines.common.distributions import make_proba_dist_type
 from stable_baselines.common.input import observation_input
 
+import warnings
```
Minor: please put built-in import first.
@hill-a we should write those rules somewhere, so for imports order:
1. built-in
2. packages
3. current module

with one space between each,  that gives for instance:
```python
import os
import warnings

import numpy as np

from stable_baselines import PPO2
```

araffin commented on stable_baselines/common/input.py:
```diff
@@ -17,33 +17,34 @@ def observation_input(ob_space, batch_size=None, name='Ob', scale=False):
-        input_x = tf.placeholder(shape=(batch_size,), dtype=tf.int32, name=name)
-        processed_x = tf.to_float(tf.one_hot(input_x, ob_space.n))
-        return input_x, processed_x
+        observation_ph = tf.placeholder(shape=(batch_size,), dtype=tf.int32, name=name)
```
I like that renaming, good initiative ;)

araffin commented on stable_baselines/common/policies.py:
```diff
@@ -25,6 +27,65 @@ def nature_cnn(scaled_images, **kwargs):
+def mlp_extractor(flat_observations, net_arch, act_fun):
+    """
+    Constructs an MLP that receives observations as an input and outputs a latent representation for the policy and
+    a value network. The `layers_spec` parameter allows to specify the amount and size of the hidden layers and how many
```
```suggestion
    a value network. The ``net_arch`` parameter allows to specify the amount and size of the hidden layers and how many
```

araffin commented on stable_baselines/common/policies.py:
```diff
@@ -25,6 +27,65 @@ def nature_cnn(scaled_images, **kwargs):
+
+    For example to construct a network with one shared layer of size 55 followed by two non-shared layers for the value
+    network of size 255 and a single non-shared layer of size 128 for the policy network, the following layers_spec
+    would be used: `[55, dict(vf=[255, 255], pi=[128])]`. A simple shared network topology with two layers of size 128
```
```suggestion
    would be used: ``[55, dict(vf=[255, 255], pi=[128])]``. A simple shared network topology with two layers of size 128
```

araffin commented on stable_baselines/common/policies.py:
```diff
@@ -25,6 +27,65 @@ def nature_cnn(scaled_images, **kwargs):
+    would be used: `[55, dict(vf=[255, 255], pi=[128])]`. A simple shared network topology with two layers of size 128
+    would be specified as [128, 128].
+
+    :param flat_observations: The observations to base policy and value function on.
```
```suggestion
    :param flat_observations: (tf.Tensor) The observations to base policy and value function on.
```
Please specify the type for the others inputs

araffin commented on tests/test_custom_policy.py:
```diff
@@ -12,7 +12,7 @@
     def __init__(self, *args, **kwargs):
         super(CustomCommonPolicy, self).__init__(*args, **kwargs,
-                                           layers=[8, 8],
+                                           net_arch=[dict(vf=[8, 8], pi=[8, 8])],
```
```suggestion
                                           net_arch=[8, dict(vf=[8, 8], pi=[8, 8])],
```
That way you can test the shared option too ;)

araffin commented:
@hill-a can't we rather fix the save function?

araffin commented:
@hill-a @erniejunior yes, that is dirty but that is the only way to be backward compatible as params are saved only in a list (no name associated with it)
And for now, I would say we should avoid breaking people's code.

erniejunior commented:
@hill-a I implemented your suggestion using `zip_longest` from the itertools. This makes it slightly more readable. All other requests were resolved I think. Now lets wait for the CI.
@hill-a @araffin thank you for the valuable feedback and debugging the loading issue!

erniejunior commented:
Great!

If it makes matters simpler, I am happy to become a maintainer!

araffin commented:
@erniejunior cool =), welcome to the team then ;)

hill-a commented:
Invitation sent, Merging PR
